File: server/src/models/exam.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import TYPE_CHECKING, Sequence

if TYPE_CHECKING:
    from agents.exam.exam_agent import ExamAgent


@dataclass(slots=True)
class ExamQuestion:
    """Модель одного экзаменационного вопроса."""

    id: int
    text: str
    reference_answer: str | None = None


@dataclass(slots=True)
class ExamResult:
    """Результат проверки ответа студента."""

    question: ExamQuestion
    user_answer: str
    verdict: str
    recommendation: str
    issues: list[str] = field(default_factory=list)
    score: str | None = None
    raw_feedback: str | None = None


class ExamSession:
    """Помощник, который проводит студента через экзамен."""

    def __init__(
        self,
        agent: ExamAgent,
        *,
        questions: Sequence[ExamQuestion],
        context: str,
        language: str = "ru",
    ) -> None:
        self._agent = agent
        self._context = context
        self._language = language
        self._questions = list(questions)
        self._cursor = 0
        self._active_question: ExamQuestion | None = None
        self.results: list[ExamResult] = []

    def has_next_question(self) -> bool:
        """Возвращает True, если в экзамене есть еще не заданные вопросы."""
        return self._cursor < len(self._questions)

    def next_question(self) -> ExamQuestion:
        """Возвращает следующий вопрос экзамена или выбрасывает StopIteration."""
        if not self.has_next_question():
            raise StopIteration("В этом экзамене больше нет вопросов.")
        question = self._questions[self._cursor]
        self._cursor += 1
        self._active_question = question
        return question

    def submit_answer(self, answer: str) -> ExamResult:
        """Оценивает ответ студента на активный вопрос и сбрасывает активный вопрос."""
        if self._active_question is None:
            raise ValueError("Нет активного вопроса для оценки.")
        result = self._agent.evaluate_answer(
            question=self._active_question,
            answer=answer,
            context=self._context,
            language=self._language,
        )
        self.results.append(result)
        self._active_question = None
        return result

File: client/src/types/exam.ts
/** Вопрос экзамена (POST /courses/{course_id}/exam/start) */
export interface ExamQuestion {
  id: number;
  text: string;
  reference_answer: string | null;
}

/** Ответ на старт экзамена */
export interface ExamStartResponse {
  exam_id: string;
  question: ExamQuestion;
  has_more_questions: boolean;
}

/** Оценка ответа студента (POST /courses/{course_id}/exam/answer) */
export interface ExamEvaluation {
  question_id: number;
  question_text: string;
  user_answer: string;
  verdict: string;
  recommendation: string;
  issues: string[];
  score: string;
  raw_feedback: string;
}

/** Ответ на ответ студента */
export interface ExamAnswerResponse {
  evaluation: ExamEvaluation;
  next_question: ExamQuestion | null;
  has_more_questions: boolean;
  completed: boolean;
}

export interface ExamResultItem {
  question_id: number;
  question_text: string;
  user_answer: string;
  verdict: string;
  recommendation: string;
  issues: string[];
  score: string;
  raw_feedback: string;
}

/** Сводка по экзамену (GET /courses/{course_id}/summary) */
export interface ExamSummaryResponse {
  exam_id: string;
  course_id: number;
  completed: boolean;
  pending_question: ExamQuestion | null;
  results: ExamResultItem[];
}
